#!/usr/bin/env python
# chmod a+x streamer.cfg
PACKAGE = "thermalvis"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

detector_datatype = gen.enum([ 
	gen.const("RAW", int_t, 0, "Raw detector mode"),
	gen.const("LUM", int_t, 1, "Luminance-corrected mode"),
	gen.const("INS", int_t, 2, "Instrument mode (recommended)"),
	gen.const("RAD", int_t, 3, "Radiance mode (radiometric cameras only)"),
	gen.const("TMP", int_t, 4, "Temperature mode (radiometric cameras only)") ],
	"An enum to set the detector output mode")
	
gen.add("detectorMode", int_t, 0, "Output of thermal-infrared detector", 2, 0, 4, edit_method=detector_datatype)

usb_datatype = gen.enum([ 
	gen.const("16", int_t, 1, "Camera will output in 16-bit format"),
	gen.const("8", int_t, 2, "Camera will output in 8-bit format"), ],
	"An enum to set the usb output mode")
	
gen.add("usbMode", int_t, 0, "USB output format (16/8-bit) of camera", 1, 1, 2, edit_method=usb_datatype)

input_datatype = gen.enum([ 
	gen.const("8bit", int_t, 0, "8-bit single channel"),
	gen.const("raw", int_t, 1, "Raw 16-bit format"),
	gen.const("mm", int_t, 2, "8-bit dual channel (multimodal)"),
	gen.const("16uc1", int_t, 3, "16-bit from kinect depth") ],
	"An enum to set the bits/pixel")
	
gen.add("inputDatatype", int_t, 0, "Pixel format of input", 1, 0, 3, edit_method=input_datatype)

gen.add("maxReadAttempts", int_t, 0, "Maximum number of times to attempt to read frame before accepting failure (0 = infinite)", 0, 0, 100)

gen.add("framerate", double_t, 0, "Desired Framerate", -1.0, -1.0, 100.0)

norm_enum = gen.enum([ 
	gen.const("standard", int_t, 0, "Standard dynamic range"),
	gen.const("equalize", int_t, 1, "Equalization"),
	gen.const("clahe", int_t, 2, "CLAHE"),
	gen.const("clahe_adaptive", int_t, 3, "Adaptive Contrast Enhancement"),
	gen.const("centralized", int_t, 4, "Centers intensities around median"),
	gen.const("expanded", int_t, 5, "Centers intensities around mid-range") ],
	"An enum to set the normalization scheme")
	
gen.add("normMode", int_t, 0, "Histogram Normalization Mode", 0, 0, 5, edit_method=norm_enum)

gen.add("normFactor", double_t, 0, "Normalization factor", 0.0, 0.0, 1.0)

filter_enum = gen.enum([ 
	gen.const("none", int_t, 0, "No filtering"),
	gen.const("gaussian", int_t, 1, "Gaussian Filter"),
	gen.const("bilateral", int_t, 2, "Bilateral Filter"),
	gen.const("adaptive_bilateral", int_t, 3, "Adaptive Bilateral Filter")],
	"An enum to set the normalization scheme")

gen.add("output16bit", bool_t, 0, "Activate 16-bit (raw) output", False)
gen.add("output8bit", bool_t, 0, "Activate 8-bit (mono) output", False)
gen.add("outputColor", bool_t, 0, "Activate color output", True)

mapping_enum = gen.enum([ 
	gen.const("GRAYSCALE", int_t, 0, "Grayscale coloring"),
	gen.const("CIECOMP", int_t, 1, "CIECOMP coloring"),
	gen.const("BLACKBODY", int_t, 2, "Blackbody coloring"),
	gen.const("RAINBOW", int_t, 3, "Rainbow coloring"),
	gen.const("IRON", int_t, 4, "Iron coloring"),
	gen.const("BLUERED", int_t, 5, "Blue-red coloring"),
	gen.const("JET", int_t, 6, "Jet coloring"),
	gen.const("CIELUV", int_t, 7, "CIELUV coloring"),
	gen.const("ICEIRON", int_t, 8, "Ice and iron coloring"),
	gen.const("ICEFIRE", int_t, 9, "Ice and fire coloring"),
	gen.const("REPEATED", int_t, 10, "Ice and fire coloring"),
	gen.const("HIGHLIGHTED", int_t, 11, "Highlights dominant pixels as red") ],
	"An enum to set the color mapping scheme")
                  
gen.add("map", int_t, 0, "A color mapping parameter which is edited via an enum", 7, 0, 11, edit_method=mapping_enum)

type_enum = gen.enum([ 
	gen.const("CV_8UC1", int_t, 0, "8-bit single channel"),
	gen.const("CV_8UC3", int_t, 1, "8-bit triple channel"),
	gen.const("CV_16UC1", int_t, 2, "16-bit single channel") ],
	"An enum to set the output image type")
	
gen.add("wantsToUndistort", bool_t, 0, "Activate image undistortion", False)

gen.add("fusionFactor", double_t, 0, "Scaling range for modality fusion", 0.6, 0.0, 1.0)

gen.add("serialPollingRate", double_t, 0, "Rate (per sec) for serial polling", 25.0, 0.0, 100.0)
gen.add("maxNucInterval", int_t, 0, "Maximum NUC interval (in sec)", 45, 0, 300)
gen.add("maxNucThreshold", double_t, 0, "Maximum NUC threshold (in deg C)", 0.2, 0.0, 100.0)

gen.add("verboseMode", bool_t, 0, "Display additional debug info", False)
gen.add("autoTemperature", bool_t, 0, "Automatically select temperature limits based on image range", False)
gen.add("minTemperature", double_t, 0, "Minimum temperature", 20.0, -50.0, 200.0)
gen.add("maxTemperature", double_t, 0, "Maximum temperature", 40.0, -50.0, 200.0)

gen.add("degreesPerGraylevel", double_t, 0, "Temperature quantity represented by a single input graylevel", 0.01, 0.00, 0.20)
gen.add("desiredDegreesPerGraylevel", double_t, 0, "Temperature quantity represented by a single output graylevel", 0.05, 0.00, 0.20)
gen.add("zeroDegreesOffset", int_t, 0, "Input graylevel value of temperature of zero degrees", 0, 0, 65535)
gen.add("threshFactor", double_t, 0, "Proportion of graylevels to threshold", 0.0, 0.0, 1.0)

exit(gen.generate(PACKAGE, "streamer", "streamer"))
